pkg/ring/basic_lifecycler.go

```diff
@@ -0,0 +1,474 @@
+package ring
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cortexproject/cortex/pkg/ring/kv"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+type BasicLifecyclerDelegate interface {
```

Comment:
Some naming nitpicking... `OnRingInstance` prefix can perhaps be reduced to `On`?

Reply:
It's quite verbose right now. The reason why I added the prefix `OnRingInstance` is because then you can add it directly to a service struct (ie. `StoreGateway`) with no ambiguity. I'm open to reconsider it. Thoughts?

Reply:
I don't see problem with shorter names, but this works too.